mkosi/backend.py

```diff
@@ -950,3 +952,44 @@ def complete_step(cls, text: str, text2: Optional[str] = None) -> Iterator[List[
 
         if text2 is not None:
             cls.print_step(text2.format(*args))
+
+def path_startswith_any(p: Path, prefixes: Iterable[str]) -> bool:
+    for prefix in prefixes:
+        if str(p).startswith(prefix):
+            return True
+
+    return False
+
+
+def log_umount_blockers(mount: Path) -> None:
+    blockers: List[Tuple[str, Path]] = []
+
+    for d in Path("/proc").iterdir():
+        if not d.is_dir():
+            continue
+
+        try:
+            int(d.name)
+        except ValueError:
+            continue
+
+        if not d.joinpath("fd").exists():
+            continue
+
+        comm = d.joinpath("comm").read_text()
+
+        for p in d.joinpath("fd").iterdir():
+            p = p.resolve()
+
+            if not str(p).startswith(str(mount)):
+                continue
+
+            if path_startswith_any(p, ("/dev", "/sys", "/proc")):
+                continue
+
+            blockers += [(comm, p)]
+
+    if blockers:
+        MkosiPrinter.warn(f"Unmounting '{mount}' is blocked by:")
+        for comm, p in blockers:
+            MkosiPrinter.warn(f"- {comm} ({p})")
```

Comment:
```suggestion
            MkosiPrinter.warn(f"- {comm} ({p})")
```
and the `MkosiPrinter.warn(f"Unmounting '{mount}' is blocked by:")` can be moved to any point before the beginning of the loop, since in we are in this function we know we are being blocked by *something* already.

Reply:
Not entirely. umount can fail for a number of reasons, and unfortunately we can't detect whether it's EBUSY or something else so it could potentially be that there's a different error and that there aren't any processes blocking the unmount. We could grep the error message but that's kinda nasty, so this seemed like the better option.

Reply:
True, there could be other reasons for umount failure, didn't think about that. TIL, only has an opaque 1 return code on error. How helpful.

I still think printing the blocking files right away, maybe as `{comm} blocking on {p}` or `{comm} holding {p} open` would be simpler than collecting them all first, but at this point that's a question of stylistic preference.